llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -2034,6 +2080,202 @@ PackOp::reifyResultShapes(OpBuilder &builder,
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// UnPackOp
+//===----------------------------------------------------------------------===//
+
+// Get the tile sizes as `OpFoldResult`.
+SmallVector<OpFoldResult> UnPackOp::getMixedTiles() {
+  return ::getMixedTiles(*this);
+}
+
+// Return a mapping from positions `dims_pos` to their `OpFoldResult` tile
+// factors.
+DenseMap<int64_t, OpFoldResult> UnPackOp::getDimAndTileMapping() {
+  return ::getDimAndTileMapping(*this);
+}
+
+// Implement the tiling interface. Generate the scalar implementation for the
+// `unpack` operation.
+LogicalResult UnPackOp::generateScalarImplementation(OpBuilder &builder,
+                                                     Location loc,
+                                                     ValueRange ivs) {
+  assert(ivs.size() == getOutputRank() &&
+         "number of ivs must match the rank of the output tensor");
+  OpBuilder::InsertionGuard g(builder);
+  // The `ivs` already represent the position into the output tensor.
+  ReifiedRankedShapedTypeDims outputShape;
+  if (failed(reifyResultShapes(builder, outputShape)))
+    return getOperation()->emitOpError("failed to reify result shape");
+  if (outputShape.size() != 1 || outputShape[0].size() != getOutputRank()) {
+    return getOperation()->emitOpError(
+               "expected shape of one result value of rank")
+           << getOutputRank();
+  }
+
+  DenseMap<int64_t, OpFoldResult> dimAndTileMapping = getDimAndTileMapping();
+  // untiled loops and tile loops induction variables.
+  SmallVector<Value> sourceIvs;
+  // point loops induction variables.
+  SmallVector<Value> sourceIvsPointLoops;
+  sourceIvs.reserve(getOutputRank());
+  sourceIvsPointLoops.reserve(dimAndTileMapping.size());
+  for (auto dim : llvm::seq<int64_t>(0, getOutputRank())) {
+    if (dimAndTileMapping.count(dim)) {
+      DivModValue divMod =
+          getDivMod(builder, loc, ivs[dim],
+                    getAsValues(builder, loc, dimAndTileMapping[dim])[0]);
```

Comment:
This can be `getValueOrCreateConstantIndexOp`.

```suggestion
                    getValueOrCreateConstantIndexOp(builder, loc, dimAndTileMapping[dim]));
```